docs/architecture/adt-014-proportional-slashing.md

```diff
@@ -0,0 +1,78 @@
+
+We can achieve this by not only taking into account the sum of the percentages of the validators that faulted, but also the *number* of validators that faulted in the window.  One general form for an equation that fits this desired property looks like this:
+
+```
+slash_amount = k * (sqrt(power_1) + sqrt(power_2) + ... + sqrt(power_n))^2
+```
+
+So now, for example, (assuming k = 1) if one validator of 10% faults, it gets a 10% slash, while if two validators of 5% each fault together, they both get a 20% slash ((sqrt(0.05)+sqrt(0.05))^2).
+
+The k constant can be provided by the governable parameters, to for example, punish double signs more heavily than liveness faults.  The k factor can also be something other than a constant, there is some research on using things like inverse gini coefficients to mitigate some griefing attacks, but this an area for future research.
+
+One will note, that this model doesn't differentiate between multiple validators run by the same operators vs validators run by different operators.  This can be seen as an additional benefit in fact.  It incentivizes validators to differentiate their setups from other validators, to avoid having correlated faults with them or else they risk a higher slash.  So for example, operators should avoid using the same popular cloud hosting platforms or using the same Staking as a Service providers.  This will lead to a more resilient and decentralized network.
+
+### Implementation
+
+In the slashing module, we will add two queues that will track all of the recent slash events.  For double sign faults, we will define "recent slashes" as ones that have occured within the last `unbonding period`.  For liveness faults, we will define "recent slashes" as ones that have occured withing the last `jail period`.
+
+```
+type SlashEvent struct {
+    Address                 sdk.ValAddress
+    ValidatorVotingPercent  sdk.Dec
+    SlashedSoFar            sdk.Dec
+}
+```
+
+These slash events will be pruned from the queue once they are older than their respective "recent slash period".
+
+Whenever a new slash occurs, a `SlashEvent` struct is created with the faulting validator's voting percent and a `SlashedSoFar` of 0.  Because recent slash events are pruned before the unbonding period and unjail period expires, it should not be possible for the same validator to have multiple SlashEvents in the same Queue at the same time.
+
+We then will iterate over all the SlashEvents in the queue, using their `ValidatorVotingPercent` to calculate the new percent to slash all the validators in the queue at, using the "Square of Sum of Roots" formula introduced above.
+
+Once we have the `NewSlashPercent`, we then iterate over all the `SlashEvent`s in the queue once again, and if `NewSlashPercent > SlashedSoFar` for that SlashEvent, we call the `staking.Slash(slashEvent.Address, NewSlashPercent - SlashedSoFar)`.  We then set `SlashEvent.SlashedSoFar` amount to `NewSlashPercent`.
```

Comment:
From my understanding this requires 2 iterations over the `SlashQueue` every time a new slash occurs. Is that correct?

Also since these validators are getting slashed incrementally, I'm not sure this solution is correct:

Let's say we want to slash a Validator A by 75% cumulatively and it initially has 100 ATOM.

Say in Round 1, the `NewSlashPercent` is 50%:
ValidatorA now has 50ATOM

Say in Round 2, the `NewSlashPercent` is 75%"
So Validator will now get slashed `75-50 = 25%`:
ValidatorA now has 37.5 ATOM

If the Slash of 75% happened at once: The Validator will have 25ATOM

So there is a difference here which causes incorrect behavior from my understanding. Let me know if my thinking is wrong here

Reply:
In the stakingkeeper Slash function, you pass in what power that the slash percent is based on.  And so we pass in the original power, which we stored in the `SlashEvent` struct.

So in your example, when we call slash in Round 2, we say slash 25% of the original power of 100 Atoms.

See implementation here:  https://github.com/cosmos/cosmos-sdk/blob/sunny/proportional-slashing/x/slashing/internal/keeper/infractions.go#L249

Reply:
I'll add this note to the ADR spec.